### PR TITLE
[Dubbo-4587]registry config duplicate process, registered url remove timestamp key

### LINKED ISSUE #4587: 重复初始化consumer时，AbstractRegistry.register(URL url)由于url中的timestamp key始终不同，导致一直重复添加，直到内存溢出
- [x] I have searched the [issues](https://github.com/apache/dubbo/issues) of this repository and believe that this is not a duplicate.
- [x] I have checked the [FAQ](https://github.com/apache/dubbo/blob/master/FAQ.md) of this repository and believe that this is not a duplicate.

### Environment

* Dubbo version: 2.5.3
* Operating System version: mac
* Java version: 1.8.0_202

### Steps to reproduce this issue

1. java编码方式重复初始化一个没有服务提供者的consumer
```
while (true) {
            try {

                ReferenceConfig<INoProvider> reference = new ReferenceConfig<>(); // #1
                reference.setApplication(new ApplicationConfig("first-dubbo-client")); // #2
                reference.setRegistry(new RegistryConfig("zookeeper://127.0.0.1:2181")); // #3
                reference.setInterface(INoProvider.class); // #4
                INoProvider noProvider = reference.get(); // #5
                noProvider.say(); // #6

                Thread.sleep(10);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
```

AbstractRegistry方法

```
public void register(URL url) {
        if (url == null) {
            throw new IllegalArgumentException("register url == null");
        }
        if (logger.isInfoEnabled()){
            logger.info("Register: " + url);
        }
        registered.add(url);
    }

```
AbstractRegistry内部的registered，subscribed，notified，都有重复添加的情况，因为URL的参数中带有timestamp key，每次的timestamp不同，URL的hashCode()方法并没有排除timestamp，导致能一直重复add

### PR DESCRIPTION
**Pull request by tswstarplanet**
## What is the purpose of the change

1.fix #4587
2.registry config duplicate process
3.split the address of RegistryConfig when use zookeeper as the default config-center
4.dubbo-api-demo pom.xml add zookeeper registry and config-center dependency

## Brief changelog

1:
AbstractInterfaceConfig.java
ServiceConfig.java
ReferenceConfig.java
RegistryProtocol.java
RegistryProtocolTest.java

2.AbstractInterfaceConfig.java  loadRegistries(boolean provider) method

3.AbstractInterfaceConfig.java  useRegistryForConfigIfNecessary(URL url) method

4.pom.xml of dubbo-api-demo
## Verifying this change

XXXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Comment by tswstarplanet** [comment]
@beiwei30 Would you please review this pr ?

**Comment by guohao** [comment]
@tswstarplanet Could you fix the conflicts?

File: dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/integration/RegistryProtocol.java [hunk 1]
```diff
@@ -122,8 +122,8 @@
 
     private final static Logger logger = LoggerFactory.getLogger(RegistryProtocol.class);
     private static RegistryProtocol INSTANCE;
-    private final Map<URL, NotifyListener> overrideListeners = new ConcurrentHashMap<>();
-    private final Map<String, ServiceConfigurationListener> serviceConfigurationListeners = new ConcurrentHashMap<>();
+    private final Map<URL, List<NotifyListener>> overrideListeners = new ConcurrentHashMap<>();
```
Comment (beiwei30) [inline dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/integration/RegistryProtocol.java:125]: I am not convinced that we need a list of overrideListeners and ServiceConfigurationListener here. I will leave this to @chickenlj to judge.

Comment (chickenlj) [reply dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/integration/RegistryProtocol.java:125]: what's the difference between `serviceKey` classified listeners and `URL` classified listeners?
 Map<String, ServiceConfigurationListener> serviceConfigurationListeners 
Map<URL, List<NotifyListener>> overrideListeners

**Pull request closed without merging**